### PR TITLE
Add support for AllVersionsExpiration ilm rule

### PR DESCRIPTION
**Pull request by dhananjaykrutika**
This is a minio-only rule for removing all versions of an object upon expiration. The rule takes two parameters: Days and DeleteMarker.
"Days" indicates the number of days after which the object and all its versions must be expired.
The optional "DeleteMarker" flag indicates that the expiration be applied if the latest version of this object is a delete marker.

These specific changes in minio-go are for consumption by ec in a different PR.

### CODE REVIEW

File: pkg/lifecycle/lifecycle.go [hunk 1]
```diff
@@ -434,12 +434,37 @@ func (de DelMarkerExpiration) MarshalXML(enc *xml.Encoder, start xml.StartElemen
...
+	Days         int                `xml:"Days,omitempty" json:"Days,omitempty"`
+	DeleteMarker ExpireDeleteMarker `xml:"DeleteMarker,omitempty" json:"DeleteMarker,omitempty"`
+}
+
+// IsDaysNull returns true if days field is null
+func (e AllVersionsExpiration) IsDaysNull() bool {
+	return e.Days == 0
+}
+
+func (e AllVersionsExpiration) IsNull() bool {
```
Comment (klauspost) [inline pkg/lifecycle/lifecycle.go:449]: Do we need both?

File: pkg/lifecycle/lifecycle.go [hunk 2]
```diff
@@ -434,12 +434,37 @@ func (de DelMarkerExpiration) MarshalXML(enc *xml.Encoder, start xml.StartElemen
...
 }
 
+// AllVersionsExpiration represents AllVersionsExpiration actions element in an ILM policy
+type AllVersionsExpiration struct {
+	XMLName      xml.Name           `xml:"AllVersionsExpiration" json:"-"`
+	Days         int                `xml:"Days,omitempty" json:"Days,omitempty"`
+	DeleteMarker ExpireDeleteMarker `xml:"DeleteMarker,omitempty" json:"DeleteMarker,omitempty"`
+}
+
+// IsDaysNull returns true if days field is null
```
Comment (klauspost) [inline pkg/lifecycle/lifecycle.go:444]:
```suggestion
// IsDaysNull returns true if days field is 0.
```

File: pkg/lifecycle/lifecycle.go [hunk 3]
```diff
@@ -434,12 +434,37 @@ func (de DelMarkerExpiration) MarshalXML(enc *xml.Encoder, start xml.StartElemen
...
+func (e AllVersionsExpiration) IsDaysNull() bool {
+	return e.Days == 0
+}
+
+func (e AllVersionsExpiration) IsNull() bool {
+	return e.IsDaysNull()
+}
+
+func (ave AllVersionsExpiration) MarshalXML(enc *xml.Encoder, start xml.StartElement) error {
```
Comment (klauspost) [inline pkg/lifecycle/lifecycle.go:453]:
```suggestion
// MarshalXML satisfies xml.Marshaler to provide custom encoding.
func (e AllVersionsExpiration) MarshalXML(enc *xml.Encoder, start xml.StartElement) error {
```
ave -> e below.

File: pkg/lifecycle/lifecycle.go (hunk 1 above)
Comment (klauspost) [inline pkg/lifecycle/lifecycle.go:449]:
```suggestion
// IsNull returns whether the expiration is enabled.
func (e AllVersionsExpiration) IsNull() bool {
```

File: pkg/lifecycle/lifecycle.go [hunk 4]
```diff
@@ -434,12 +434,37 @@ func (de DelMarkerExpiration) MarshalXML(enc *xml.Encoder, start xml.StartElemen
...
+	if ave.IsNull() {
+		return nil
+	}
+	type allVersionsExp AllVersionsExpiration
+	return enc.EncodeElement(allVersionsExp(ave), start)
```
Comment (klauspost) [inline pkg/lifecycle/lifecycle.go:458]: Why is this cast needed?

**Review by klauspost — COMMENTED** [review]
A few minor things

File: pkg/lifecycle/lifecycle.go (hunk 4 above)
Comment (klauspost) [reply pkg/lifecycle/lifecycle.go:458]: Ah, to avoid the recursive call 👍🏼

**Review by klauspost — APPROVED** [review]
LGTM

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**